### PR TITLE
checkstyle: ensure overloaded methods are grouped together #229

### LINKED ISSUE #229: checkstyle: ensure overloaded methods are grouped together
This improves the readability of the code as it allows developers to see all possible variations of a method at one place.

### PR DESCRIPTION
**Pull request by chao1995**
Fixes #229 

Ready for review.

### CODE REVIEW

**Comment by damithc** [comment]
@chao1995 you can request a peer review from one of the team member you think is suitable to review this.

**Comment by damithc** [comment]
Also, you can add the label `toReview` :-)

**Comment by yamgent** [comment]
I tested it and it seems to be working most of the time.

However, I was able to break this rule if there are other class/interface declarations in the middle of the two methods. Not sure whether we can fix this or not (and whether we should be concerned about it):

```diff
@@ -101,14 +101,18 @@ public class ModelManager extends ComponentManager implements Model {
    private void updateFilteredPersonList(Expression expression) {
        filteredPersons.setPredicate(expression::satisfies);
    }

    //========== Inner classes/interfaces used for filtering =================================================

    interface Expression {
        boolean satisfies(ReadOnlyPerson person);
        String toString();
    }

+   public void updateFilteredPersonList(String keyword, boolean someRandomBool) {
+       // do nothing
+   }
+
    private class PredicateExpression implements Expression {

        private final Qualifier qualifier;
```

**Comment by damithc** [comment]
> However, I was able to break this rule if there are other class/interface declarations in the middle of the two methods. Not sure whether we can fix this or not:

I think we don't need to be worried about it. I sounds like a loophole in checkstyle which could be fixed in a future version.

**Comment by damithc** [comment]
@m133225 for your inputs

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -195,7 +195,7 @@
 
 
     <!--
-    MODIFIERS CHECKS
+    ORDER CHECKS
```
Comment (damithc) [inline config/checkstyle/checkstyle.xml]: What's the reason for this change?

Comment (chao1995) [inline config/checkstyle/checkstyle.xml]: This is because both "ModifierOrder" and "OverloadMethodsDeclarationOrder" have "Order" as suffixes.

Comment (damithc) [inline config/checkstyle/checkstyle.xml]: I see. I wonder if these terms were in the original default checkstyle configuration that came with checkstyle (if there is such a thing) and in that case we may want to keep the same.

Comment (chao1995) [inline config/checkstyle/checkstyle.xml]: Yes both "ModifierOrder" and "OverloadMethodsDeclarationOrder" are directly provided by checkstyle. So I think we are using them right (without modification to the names).

**Comment by chao1995** [comment]
@yamgent Nice spot! Thanks!

As @damithc said, this is a potential loophole in checkstyle, though we can add ["InnerTypeLast"](http://checkstyle.sourceforge.net/config_design.html#InnerTypeLast) rule to help this specific case, so the violating overloaded method should not appear after the inner interface.

However, "InnerTypeLast" is not specified in Google Java code style, so I think for this PR's concern, we just don't add the "InnerTypeLast" first (maybe we can add in a separate PR if necessary), but add more info in the commit message documenting this potential loophole in checkstyle.

**Comment by chao1995** [comment]
@m133225 @yamgent Thanks for the review! Updated commit message.

**Comment by damithc** [comment]
@chao1995 can rebase branch while waiting for @m133225 's approval?

**Comment by chao1995** [comment]
@damithc Rebased.

**Comment by pyokagan** [comment]
Some comments on the commit message:

> Check overloaded methods are grouped together in checkstyle
> 

The rationale for this change is missing. Why do we need this extra checkstyle rule?

> Note that there is a potential loophole in checkstyle
> OverloadMethodsDeclarationOrder module which permits out-of-order
> overloaded methods exist when they are separated by a nested class

This reads a bit strangely. Maybe remove `exist`?

> or interface.
> 
> e.g. the overloaded updateFilteredPersonList in ModelManager.java
> will pass.

I think we can rephrase this more clearly as:
```
e.g. the following code in ModelManager.java still passes the check, even
though there is a nested interface between the two `updateFilteredPersonList()`
method definitions:
```
> 
> private void updateFilteredPersonList(Expression expression) {
>     filteredPersons.setPredicate(expression::satisfies);
> }
> 
> //========== Inner classes/interfaces ==================
> interface Expression {
>     boolean satisfies(ReadOnlyPerson person);
>     String toString();
> }
> 
> public void updateFilteredPersonList(String keyword, boolean someRandomBool) {
>     // do nothing
> }

Indent the code by 4 spaces to make it clear that it is a code block.

**Comment by m133225** [comment]
@chao1995 Agree with the changes that @pyokagan has suggested. Do remember to update it!

**Comment by CanIHasReview** [comment]
# v1

@chao1995 submitted v1 for review.

* [v1 1/1] [Check overloaded methods are grouped together in checkstyle](https://github.com/se-edu/addressbook-level4/pull/251/commits/dbf8a42ea961a54ac6efbbef8a3a1e9c5d9c62a6)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/8164678f10e07049144c98acd52fc3572ad5c9a8...dbf8a42ea961a54ac6efbbef8a3a1e9c5d9c62a6))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/251/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by chao1995** [comment]
@m133225 Updated!

**Comment by damithc** [comment]
@chao1995 do a rebase  while we wait for @pyokagan to have another look?

**Comment by pyokagan** [comment]
Looks good overall.

Just a small nit that I think @chao1995 missed out from https://github.com/se-edu/addressbook-level4/pull/251#issuecomment-273757966 :

> Indent the code by 4 spaces to make it clear that it is a code block.

I've fixed that up on my end.

Also, the commit message wasn't wrapped at 72 characters. I fixed that up on my end too.
